### PR TITLE
Optional metadata fields when managing an application

### PR DESCRIPTION
**Pull request by AramAlsabti**
Related [frontend PR](https://github.com/OS2iot/OS2IoT-frontend/pull/72)

Note that typeorm was updated to make it easier to update the list of controlled properties and device types on an application. Previously, you would either need to delete the old child entities before adding the new ones or fetch and update the existing ones while inserting any new entities. Otherwise, the application id will be `null`'ed in the controlled property or device type table.

### CODE REVIEW

File: src/migration/1646669134921-application-metadata.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
+import {MigrationInterface, QueryRunner} from "typeorm";
+
+export class applicationMetadata1646669134921 implements MigrationInterface {
+    name = 'applicationMetadata1646669134921'
+
+    public async up(queryRunner: QueryRunner): Promise<void> {
+        await queryRunner.query(`ALTER TABLE "api_key_permissions_permission" DROP CONSTRAINT "FK_c1141b0748c24b2f3e78789b6c8"`);
```
Comment (GufCab) [inline src/migration/1646669134921-application-metadata.ts:7]: What's going on here? Why are all these constraints changed?

File: src/entities/base-no-id.entity.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
+import {
+    CreateDateColumn,
+    JoinColumn,
+    ManyToOne,
+    PrimaryGeneratedColumn,
+    UpdateDateColumn,
+} from "typeorm";
+
+/**
```
Comment (GufCab) [inline src/entities/base-no-id.entity.ts:9]: Why do we now need to have Composite keys?

File: src/migration/1646669134921-application-metadata.ts (hunk 1 above)
Comment (GufCab) [reply src/migration/1646669134921-application-metadata.ts:7]: If this comes from updating TypeORM - make a note as to what is going on

File: src/entities/base-no-id.entity.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ * This is similar to the base entity class but without the id column. This allows classes to use
+ * (composite) keys
+ */
+export abstract class DbBaseNoIdEntity {
```
Comment (GufCab) [inline src/entities/base-no-id.entity.ts:13]: As far as I can see, we're not using this. Remove until we need it.

**Review by GufCab — COMMENTED** [review]
Basically fine, minor comments.

File: src/entities/base-no-id.entity.ts (hunk 3 above)
Comment (AramAlsabti) [reply src/entities/base-no-id.entity.ts:13]: Yes, I forgot to delete it.

**Review by GufCab — APPROVED** [review]
Approved, merge along with the front end PR

**Pull request merged**